### ISSUE TITLE
Training from the scratch does not work

### ISSUE THREAD

**Issue by HripsimeS** [ultralytics/yolov5#9978]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### Bug

Hello. I tried to train the model from the scratch with coco128 dataset by using the command 
**python train.py --data coco128.yaml --weights '' --cfg yolov5s.yaml --img 640**

During training  these values I received **nan** results **box_loss , obj_loss, cls_loss**  

**best.pt** and **last.pt** model saved in runs\train\exp\weights. I placed the model in "weights" folder and then I tried to use **best.pt** to detect on inference mode with the following command. 
**python detect.py --weights weights/best.pt --source data/images** 

In runs\detect\exp folder I did not get predictions. I assume the training from the scratch does not work. But using Pretrained Checkpoints mentioned https://github.com/ultralytics/yolov5/releases gives good predictions. 

Can you let me know how to make training from the scratch work efficiently so when I use my trained model I can get a good prediction on inference mode with detect.py. Look forward to hearing from you and thank you in advance.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by HripsimeS** [comment]
@glenn-jocher thank you very much for all description and details. Your pretrained checkpoint  **yolov5s.pt** you trained on COCO dataset? I have a dataset of 350 images with their labels, but classes are different than those 80 clases of COCO dataset. In this case is it possible to use your pretrained checkpoint yolov5s.pt or I need to have 1500 images per class to do a training from the scratch?

**Comment by glenn-jocher** [comment]
@HripsimeS I'd recommend training from pretrained weights under all circumstances, it doesn't matter what your dataset is like, you will always get better results than training from scratch.

**Comment by HripsimeS** [comment]
@glenn-jocher thank you very much for your advice. I tried to use pretrained weights with the command below where in mydata file are my classes names and paths to the train and val data with the same structure as for COCO dataset. 

**python train.py --data mydata.yaml --weights yolov5s.pt --img 640**

I received the error below and **nan** results **box_loss , obj_loss, cls_loss**
**Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule.** 

I read that we get that error because some optimizer.step call are skipped as we use **amp** which can create invalid gradients if the loss scaling factor is too large and will thus skip the parameter updates. 

Can you please advice how I can fix the issue to not get nan values for box_loss , obj_loss, cls_loss, because for the moment it does not work even with pretrained weights. If you need additional details from my side, please let me know. 

![image](https://user-images.githubusercontent.com/42246765/199189982-a1a2ce65-dabf-4c84-8354-3730cc84213e.png)

**Comment by glenn-jocher** [comment]
@HripsimeS nan losses can be indicative of errors in your dataset or too aggressive learning rates etc causing instability. I'd update the hyps for lower LR and to increase the warmup period.

**Comment by HripsimeS** [comment]
@glenn-jocher I tried training on both coco128 and coco datasets with pretrained weights, but the same issue with nan losses.

python train.py --data coco128.yaml --weights yolov5s.pt --img 640
python train.py --data coco.yaml --weights yolov5s.pt --img 640

Did someone checked before if those two lines above are working efficiently and there are good results for losses? I was assuming that hyperparameters in train.py are at least optimized for coco dataset to work with pretrained weights.

**Comment by glenn-jocher** [comment]
@HripsimeS COCO training from scratch or pretrained both operate correctly. We train multiple models a day constantly and everything is working correctly. 

If you are having nan issues on default datasets then you may have environment or CUDA issues, I'd recommend our Docker image.

**Comment by glenn-jocher** [comment]
## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

**Comment by HripsimeS** [comment]
@glenn-jocher thanks for the reply. Indeed the problem was in CUDA or environment. I opened a new environment and did training with CPU. I received interesting results and you were right that with pretrained weights (even on coco dataset) the prediction was better rather with the model that I trained from the scratch. I will close this issue, thanks a lot 👍

**Comment by glenn-jocher** [comment]
Great, let us know if you find any bugs or think of any feature requests!
